core/src/main/java/org/elasticsearch/action/fieldstats/FieldStatsRequest.java

```
@@ -33,17 +40,111 @@
 
     public String[] fields() {
         return fields;
     }
 
     public void fields(String[] fields) {
+        if (fields == null) {
+            throw new IllegalArgumentException("Specified fields can't be null");
+        }
         this.fields = fields;
     }
 
+    public IndexConstraint[] indexConstraints() {
+        return indexConstraints;
+    }
+
+    public void indexConstraints(IndexConstraint[] indexConstraints) {
+        if (indexConstraints == null) {
+            throw new IllegalArgumentException("Specified indexContraints can't be null");
+        }
+        this.indexConstraints = indexConstraints;
+    }
+
+    public void source(BytesReference content) throws IOException {
+        List<IndexConstraint> indexConstraints = new ArrayList<>();
+        List<String> fields = new ArrayList<>();
+        try (XContentParser parser = XContentHelper.createParser(content)) {
+            String fieldName = null;
+            Token token = parser.nextToken();
+            assert token == Token.START_OBJECT;
+            for (token = parser.nextToken(); token != Token.END_OBJECT; token = parser.nextToken()) {
+                switch (token) {
+                    case FIELD_NAME:
+                        fieldName = parser.currentName();
+                        break;
+                    case START_OBJECT:
+                        if ("index_constraints".equals(fieldName)) {
+                            parseIndexContraints(indexConstraints, parser);
+                        } else {
+                            throw new IllegalArgumentException("Unknown field [" + fieldName + "]");
+                        }
+                        break;
+                    case START_ARRAY:
+                        if ("fields".equals(fieldName)) {
+                            while ((token = parser.nextToken()) != XContentParser.Token.END_ARRAY) {
+                                if (token.isValue()) {
```

does this mean we will silently ignore other tokens?